contracts/bots/PartialLiquidationBot.sol

```diff
@@ -0,0 +1,158 @@
+contract PartialLiquidationBot is IPartialLiquidationBot {
+    /// @notice Performs a partial liquidation by swapping some collateral asset for underlying with a discount
+    /// @dev Returns the amount of underlying charged and amount of assetOut received
+    /// @param params A struct encoding liquidation params:
+    ///               * creditManager - Credit Manager where the liquidated CA currently resides
+    ///               * creditAccount - Credit Account to liquidate
+    ///               * assetOut - asset that the liquidator wishes to receive
+    ///               * amountOut - amount of the asset the liquidator wants to receive
+    ///               * repay - whether to repay debt after swapping into underlying
+    ///               * priceUpdates - data for price feeds to update before liquidation
+    function liquidatePartialSingleAsset(LiquidationParams memory params) external returns (uint256, uint256) {
+        address priceOracle = ICreditManagerV3(params.creditManager).priceOracle();
+
+        /// Since we are computing debt and collateral before liquidation,
+        /// we need to update prices beforehand, if needed
+        _applyPriceUpdates(params.priceUpdates, priceOracle);
+
+        CollateralDebtData memory cdd = ICreditManagerV3(params.creditManager).calcDebtAndCollateral(
+            params.creditAccount, CollateralCalcTask.DEBT_COLLATERAL
+        );
+
+        if (cdd.twvUSD >= cdd.totalDebtUSD) revert CreditAccountNotLiquidatableException();
+
+        address creditFacade = ICreditManagerV3(params.creditManager).creditFacade();
+
+        // It's techically possible to pass underlying as assetOut
+        // However, this would reduce the HF of the account, which will result
+        // in a revert due to failed collateral check on CM side
+        address underlying = ICreditManagerV3(params.creditManager).underlying();
+
+        // In case the liquidator wants to get as much assetOut as possible,
+        // they would pass uint256.max
+        params.amountOut = params.amountOut == type(uint256).max
+            ? IERC20(params.assetOut).balanceOf(params.creditAccount)
+            : params.amountOut;
+
+        // There is a limit imposed on the maximal amount of underlying that the liquidator can swap
+        // If the debt is repaid, then that limit is totalDebt - minDebt - underlyingBalance (i.e., exactly enough to reduce debt to minDebt)
+        // If the debt is not repaid, then that limit is totalDebt - underlyingBalance (i.e., exactly enough to fully liquidate the user in the future)
+
+        uint256 amountInUnderlying;
+        {
+            uint256 underlyingBalance = IERC20(underlying).balanceOf(params.creditAccount);
+            (uint256 minDebt,) = ICreditFacadeV3(creditFacade).debtLimits();
+
+            uint256 repayable = CreditLogic.calcTotalDebt(cdd) - (params.repay ? minDebt : 0);
```

Comment:
limiting the amount of underlying added by liquidator to `assetOut`'s quota seems a bit more reasonable to prevent liquidator from taking disabled tokens or tokens that didn't count towards collateral in general

it's a bit unfortunate though that we don't enforce all non-underlying tokens to be quoted.

Reply:
Fair point, will fix